torch/csrc/distributed/c10d/ProcessGroup.cpp

```diff
@@ -217,13 +217,32 @@ void ProcessGroup::init() {
 
 c10::intrusive_ptr<ProcessGroup::Work> ProcessGroup::broadcast(
     std::vector<at::Tensor>& tensors, const BroadcastOptions& opts) {
-  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("c10d::broadcast", "")
-      .typed<c10::intrusive_ptr<::c10d::ProcessGroup::Work>(at::TensorList,
+  // static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("c10d::broadcast", "")
+  //     .typed<c10::intrusive_ptr<::c10d::ProcessGroup::Work>(at::TensorList,
+  //         const c10::intrusive_ptr<::c10d::ProcessGroup>&, int64_t, int64_t, int64_t)>();
+  // // It's awakward to unbox the opts here and box them again in the custom C++ op.
+  // // But it's also complicated to make opts as a CustomClassHolder. Leave it as it is now.
+  // return op.call(tensors, c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(this), opts.rootRank,
+  //     opts.rootTensor, opts.timeout.count());
+
+  // This code basically is copied from the last Lazy PR.
+  // AoTAutograd has a similar problem as LazyTensor where it expects the return type to be tensor|(tensor)|[tensor].
+  // Then it wraps the returns with the fx.Proxy which can then trace the op.
+  // So here we use directly broadcast_ above instead.
+  auto& tensor = tensors[0];
+  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("c10d::broadcast_", "")
+      .typed<at::Tensor(const at::Tensor&,
           const c10::intrusive_ptr<::c10d::ProcessGroup>&, int64_t, int64_t, int64_t)>();
-  // It's awakward to unbox the opts here and box them again in the custom C++ op.
-  // But it's also complicated to make opts as a CustomClassHolder. Leave it as it is now.
-  return op.call(tensors, c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(this), opts.rootRank,
+  op.call(tensor, c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(this), opts.rootRank,
```

Comment:
i'm having trouble understanding what this change does (why it helps AotAutograd tracing)?

Reply:
Here is the reference:
https://github.com/pytorch/functorch/blob/main/functorch/_src/python_key.py#L112

TL:DR: AOT expects a node to return a tensor like object such that it can transform those at::Tensors into its own proxied tensors.

Reply:
hmm. ok, but, you still return a 'work' object, no? (from ProcessGroup::broadcast)

Reply:
Yes, the work object is constructed on the fly with the returned tensor from the dispatcher op.